### PR TITLE
133663043 incorporate reachskills volunteer opportunities

### PR DESCRIPTION
**Pull request by ZmagoD**
This is a first part of https://www.pivotaltracker.com/story/show/133663043 feature. I still need to fix unit and feature tests and add additional tests for my code.

### CODE REVIEW

**Comment by tansaku** [comment]
thanks @ZmagoD - be careful those tests all fail before they pass :-)

**Comment by tansaku** [comment]
@ZmagoD how's it going here? ready for review?

**Comment by ZmagoD** [comment]
No not yet. I will write the test during the weekend..

**Comment by ZmagoD** [comment]
@tansaku I added the features/rspec tests but I have some problems with the unit test for import_reach_skills_volunteer_opportunities.rb

File: features/volunteer_opportunities/doit_volunteer_opportunities.feature [hunk 1]
```diff
@@ -6,6 +6,7 @@ Feature: Doit volunteer opportunities
 
   Background:
     Given that the doit_volunteer_opportunities flag is enabled
+    Given that the reachskills_volunteer_opportunities flag is enabled
```
Comment (tansaku) [inline features/volunteer_opportunities/doit_volunteer_opportunities.feature]: why do we need this flag enabled before running the doit service?

**Comment by tansaku** [comment]
thanks @ZmagoD - I've replicated those errors locally:

```
Failures:

  1) ImportReachSkillsVolunteerOpportunities one page of 10 ops found checks for presence and/or creates 10 ops
     Failure/Error: expect(model_klass).to have_received(:find_or_create_by).exactly(10).times
     
       (Double :model_klass).find_or_create_by(*(any args))
           expected: 10 times with any arguments
           received: 0 times with any arguments
     # ./spec/services/import_reachskills_opportunities_spec.rb:30:in `block (3 levels) in <top (required)>'

  2) ImportReachSkillsVolunteerOpportunities one page of 10 ops found removes all reachskills ops before re-adding from reachskills api
     Failure/Error: expect(model_klass).to have_received(:find_or_create_by).exactly(10).times.ordered
     
       (Double :model_klass).find_or_create_by(*(any args))
           expected: 10 times with any arguments
           received: 0 times with any arguments
     # ./spec/services/import_reachskills_opportunities_spec.rb:37:in `block (3 levels) in <top (required)>'

```

**Comment by tansaku** [comment]
So the failure here is caused by these expectations:

```rb
expect(model_klass).to have_received(:find_or_create_by).exactly(10).times
```

not matching the method being called on the model_klass in the service:

```rb
model_klass.create(
```

the simple fix is to make the expectation for `:create` however we should think about the behaviour of the system, and why we used `find_or_create_by` in the ImportDoItVolunteerOpportunities

It's because in the process of repeatedly importing volops from the feed we don't want to import the same one again and again.  We can simplify this feature to be doing only creation, but we will probably have to deal with this replication issue at some point - happy to split that out into a separate ticket though if that's easier :-)

**Comment by ZmagoD** [comment]
@tansaku here I had refactor the reachskills importer service so that checks on already created vol_ops by the hyperlink...

**Comment by tansaku** [comment]
looking good @ZmagoD - any thoughts on the codeclimate issues being raised?

**Comment by ZmagoD** [comment]
hmm this of the class length 103 instead of 100 lines of code is so so. For the other warnings I can extract the block body in a separate method to make the code more clean.

**Comment by tansaku** [comment]
great work @ZmagoD - feel free to create refactoring tickets for any other additional work that you don't want to get into in this PR

**Comment by tansaku** [comment]
we're got a few failing tests that need cleaning up though:

```
Failing Scenarios:
cucumber -p second_try features/volunteer_opportunities/reachskills_volunteer_opportunities.feature:13 # Scenario: See a map of current reachskills volunteer opportunities
cucumber -p second_try features/volunteer_opportunities/reachskills_volunteer_opportunities.feature:17 # Scenario: See a map of current reachskills and harrow volunteer opportunities
cucumber -p second_try features/volunteer_opportunities/reachskills_volunteer_opportunities.feature:32 # Scenario: See a list of current reachskills volunteer opportunities with a link to organisation page
cucumber -p second_try features/volunteer_opportunities/reachskills_volunteer_opportunities.feature:42 # Scenario: Reachskills volunteer opportunities are opened in a new page
```

**Comment by tansaku** [comment]
although they pass for me locally when I run individually ...

**Comment by tansaku** [comment]
@ZmagoD do you get any of these fails locally? I got them in CI again ...

**Comment by ZmagoD** [comment]
No locally every test passed

**Comment by tansaku** [comment]
ugh, I guess we have a javascript timing issue :-(

**Comment by tansaku** [comment]
hi @ZmagoD sorry about this - I'll have to get it under the microscope and sort it out ...

**Comment by tansaku** [comment]
sorry no progress on this @ZmagoD :-( will have to prioritize ...

**Comment by ZmagoD** [comment]
:) I tried to with sleep method but it still throw an error in travis-ci. Locally works fine... Still there is an warning on the travis build log Google Maps API warning: NoApiKeys https://developers.google.com/maps/documentation/javascript/error-messages#no-api-keys and after that the test chashes

**Comment by tansaku** [comment]
thanks @ZmagoD - I'm not seeing that error in CI - at the moment it seems to be stuck installing:

```
Installing eventmachine 1.0.9.1 with native extensions
Installing http_parser.rb 0.6.0 with native extensions
Installing eventmachine_httpserver 0.2.1 with native extensions
```

I might roll it back to the earlier commit where we were getting a specific error ...

**Comment by tansaku** [comment]
@ZmagoD I just managed to replicate the error locally - read end of my latest blog draft for details: https://github.com/AgileVentures/av-static-website/blob/7267c909c480b4cae0ab25786cf8b5db1d4c3a1c/source/blog_articles/2017-03-21-long-running-pull-requests.html.markdown 

more soon

**Comment by tansaku** [comment]
@ZmagoD - do try syncing my update to your local branch and see if you get a local failure ...

**Comment by tansaku** [comment]
@ZmagoD - so I've analyzed this and I think it wasn't a timing issue, just a lack of the VCR caches, which meant the cuke was hitting the live ReachSkills end point, where the data has changed ... another blog post coming :-)

File: db/schema.rb [hunk 2]
```diff
@@ -156,6 +156,8 @@
     t.datetime "deleted_at"
     t.string   "address"
     t.string   "postcode"
+    t.string   "reachskills_org_name"
```
Comment (tansaku) [inline db/schema.rb]: @ZmagoD I wonder if we want to consolidate these with other fields ...

**Comment by tansaku** [comment]
@ZmagoD not sure if I posted this somewhere - here's my draft blog from yesterday on this: https://github.com/AgileVentures/av-static-website/blob/0e77de92e40d042cd57806b5d7dcc10841cec2b5/source/blog_articles/2017-03-22-debugging-by-blog.html.markdown

**Comment by tansaku** [comment]
thanks @ZmagoD - I'm trying a build restart

**Comment by tansaku** [comment]
okay @ZmagoD we have a green build finally! :-) 

We have a small drop in coverage and some codeclimate issues - are there any that you like the look of for quick fixes? Others we can add refactoring tickets ...

**Comment by ZmagoD** [comment]
hi @tansaku yes probably we should add the refactoring tickets on this ...

**Comment by tansaku** [comment]
@ZmagoD great - can you fix any of the codeclimate issues you fancy, and add the rest to pivotal as refactoring tickets - then we can merge this - thanks

**Comment by tansaku** [comment]
@ZmagoD did you make the refactoring tickets? - if so, maybe post links in here for reference? many thanks

**Comment by ZmagoD** [comment]
@tansaku sorry for late response here I have open a task in PT https://www.pivotaltracker.com/story/show/144123041

**Pull request merged**